### PR TITLE
Implement physical fling animation

### PR DESCRIPTION
**Pull request by Gustl22**
Hi,

* implemented physical fling
* reduce scroll velocity for desktop platforms (as mouse is much faster, especially with phys. fling). C.f. Google Maps, and [comments ](https://github.com/mapsforge/mapsforge/issues/688)in _mapsforge_' fling animation.
* `MapEventLayer2` for testing purposes (will work only on touch devices)
    * implemented fling rotation
    * implemented fling scale

`MapEventLayer` uses new fling animation when scrolling, so we should be careful while merging.
New features highly enhance user experience on mobile devices.
If ready can "merge" them in `MapEventLayer`, too.

Used part of Androids `FlingAnimation`: the `DragForce` (see code).

If you have any concerns, let me know (e.g. code structure, wrong specified attributes, etc.). I'm glad for every help or feedback. I'm pretty sure that all works fine, but you never know. 

EDIT by @devemux86
Can be enabled before map view initialization via:
```java
Parameters.ANIMATOR2 = true;
```

### CODE REVIEW

**Comment by devemux86** [comment]
Thanks!
_I restarted Travis build, should be ok now._

> reduce scroll velocity for desktop platforms

New implementation seems tiring on desktop, better have option for such changes.
BTW Desktop and browser should behave same with mouse(?).

> MapEventLayer2 for testing purposes (will work only on touch devices)

What do you mean, it won't be usable on other platforms?

I made several comments in the code, with same principal: new implementations better exist in their own new methods, leaving the old ones intact. So users can select what behavior they wish from the library.

At the end we should provide a flag in `Parameters` class to use the new or the old implementation (which should remain as is).

**Comment by Gustl22** [comment]
> New implementation seems tiring on desktop, better have option for such changes.

Can add a constant there, I only noticed that its kind of uncontrolled with physical fling.

> BTW Desktop and browser should behave same with mouse(?).

Yes, I thought `platform.isDesktop()` indicates all desktop environments, even the browser?!

> What do you mean, it won't be usable on other platforms?

Is usable, but not with current configuration, as the mouse or zoom wheel doesn't (or better shouldn't) trigger fling animations for zoom or rotation.

Will add a commit with your proposals... Thanks :)

File: vtm/src/org/oscim/map/Animator.java [hunk 1]
```diff
@@ -220,38 +235,86 @@ public void animateZoom(long duration, double scaleBy,
...
+        mFlingRotateForce.setValueAndVelocity(0f, angularVelocity);
+
+        animFlingStart(ANIM_ROTATE);
     }
 
-    private void animStart(float duration, int state, Easing.Type easingType) {
+    /**
+     * @param scaleVelocity the scale velocity depends on screen resolution
+     */
+    public void animateFlingZoom(float scaleVelocity, float pivotX, float pivotY) {
```
[the remark this answers is not in the document]
Comment (Gustl22) [reply vtm/src/org/oscim/map/Animator.java]: This is a new method :) 
It doesn't replace another at all.

**Comment by devemux86** [comment]
> Can add a constant there, I only noticed that its kind of uncontrolled with physical fling.

Actually on Android too, the new fling seems very sensitive, like travelling too much(?).

> Yes, I thought platform.isDesktop() indicates all desktop environments, even the browser?!

No, the platforms separation is based on OpenGL implementation.
So browser WebGL is different than actual desktop OpenGL (Windows, macOS, Linux).

> Is usable, but not with current configuration, as the mouse or zoom wheel doesn't (or better shouldn't) trigger fling animations for zoom or rotation.

Don't quite understand what you mean here, all platforms should be usable.
If a regression is introduced, should be resolved, or it's something different?

File: vtm/src/org/oscim/map/Animator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (devemux86) [reply vtm/src/org/oscim/map/Animator.java]: Sorry, probably meant the `animEaseStart` changes.

**Comment by Gustl22** [comment]
> Actually on Android too, the new fling seems very sensitive, like travelling too much(?).

Yes, I know, but as I said, I orientated on G...maps. But it makes sense: On desktop you have a mouse which is much easier and faster to move. On touch devices you have to swipe arduous to get at right position (with old fling). If you want to move more controlled can use tap and hold. But maybe can deccellerate velocity a bit.

> If a regression is introduced, should be resolved, or it's something different?

No regression is introduced anywhere... These are new features, which help rotate or scale the map especially with touch gestures. (Try NewGestureLayer on Android and rotate or scale the map with a fling move). We simply don't have these gestures on desktop.

**Comment by devemux86** [comment]
> These are new features, which help rotate or scale the map especially with touch gestures.

I see, better have such new functions in their own PR to check / focus at one feature each time. 🙂

File: vtm/src/org/oscim/map/Animator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gustl22) [reply vtm/src/org/oscim/map/Animator.java]: Sorry don't understand: Except the signature nothing has changed. (and is private, so won't affect any usings)

[the remark this answers is not in the document]
Comment (devemux86) [reply vtm/src/org/oscim/map/Animator.java]: Got confused with the many changes, that seems ok then. 🙂

**Comment by Gustl22** [comment]
Pushed a version with old functionality. I rewrote it a bit, cause ease fling animation is predictable and could use `ViewController`s `moveTo`. And uses minor variables and calculations (not each step call `moveMap`). It has the exact same result.
`ViewController`s `mPos` and `mMovePoint` now should be clearer synchronized, as the objects could have been changed while simultaneously called `moveMap`.

File: vtm/src/org/oscim/map/Animator.java [hunk 2]
```diff
@@ -238,6 +240,49 @@ public void animateZoom(long duration, double scaleBy,
...
+        velocityY = velocityY * screenFactor;
+        velocityX = clamp(velocityX, xmin, xmax);
+        velocityY = clamp(velocityY, ymin, ymax);
+        if (Float.isNaN(velocityX) || Float.isNaN(velocityY)) {
+            log.debug("fling NaN!");
+            return;
+        }
+
+        double tileScale = mStartPos.scale * Tile.SIZE;
+        Point p = new Point();
```
Comment (Gustl22) [inline vtm/src/org/oscim/map/Animator.java]: Here a final static variable is conceivable, too.

**Comment by devemux86** [comment]
Thanks!
Still are many changes which don't seem relevant to PR purpose, like fling on rotate or scale methods?

Also prefer to keep previous implementation code in use, as it's already in public and working and have the new one pluggable to be able to test by users without breaking anything.
e.g. all those changes in `updateAnimation`

File: vtm/src/org/oscim/map/ViewController.java [hunk 3]
```diff
@@ -86,12 +86,12 @@ public void setMapScreenCenter(float pivotY) {
...
      * @param my the amount of pixels to move the map vertically.
      */
-    public void moveMap(float mx, float my) {
+    public synchronized void moveMap(float mx, float my) {
         ThreadUtils.assertMainThread();
 
-        Point p = applyRotation(mx, my);
+        applyRotation(mx, my, mPos.bearing, mMovePoint);
         double tileScale = mPos.scale * Tile.SIZE;
-        moveTo(mPos.x - p.x / tileScale, mPos.y - p.y / tileScale);
```
Comment (Gustl22) [inline vtm/src/org/oscim/map/ViewController.java]: `mPos` and `p` (reference to object) could have been changed while other `moveMap` calls are done. If I interpret that right.

File: vtm/src/org/oscim/map/ViewController.java [hunk 4]
```diff
@@ -120,18 +120,24 @@ else if (mPos.y < mMinY)
...
-            mMovePoint.x = mx;
-            mMovePoint.y = my;
+    /**
+     *
+     * @param mx        the amount of pixels to move the map horizontally.
+     * @param my        the amount of pixels to move the map vertically.
+     * @param bearing   the current bearing
+     * @param movePoint the out position where to move
+     */
+    public static void applyRotation(double mx, double my, float bearing, Point movePoint) {
```
Comment (devemux86) [inline vtm/src/org/oscim/map/ViewController.java]: Could name the output point as `out` and need a null check:
```java
if (out == null)
    out = new Point();
```

File: vtm/src/org/oscim/map/Animator.java [hunk 5]
```diff
@@ -220,38 +237,129 @@ public void animateZoom(long duration, double scaleBy,
...
     }
 
+    /**
+     * Use animateEaseScroll instead
+     */
+    @Deprecated
     public void animateFling(float velocityX, float velocityY,
                              int xmin, int xmax, int ymin, int ymax) {
+        animateEaseScroll(velocityX, velocityY, xmin, xmax, ymin, ymax);
+    }
```
Comment (devemux86) [inline vtm/src/org/oscim/map/Animator.java]: No need to deprecate anything yet, it's all about choices.
Also old methods should remain "as is" for public use and add new ones separately.

**Comment by Gustl22** [comment]
How about add class `Animator2` with new functionality and keep old one? So new animations are available and could tested and improved, as we are sure old animations work fine. I only think we shouldn't loose the changes I've done (for ease fling scroll, too) 😆
And add a flag in Preferences of course...

File: vtm/src/org/oscim/map/Animator.java (hunk 5 above)
Comment (Gustl22) [reply vtm/src/org/oscim/map/Animator.java]: The method itself isn't deprecated. But the name is misleading. Isn't a fling at all! It's an ease animation for scroll/move. So I think this shouldn't used...

Comment (devemux86) [reply vtm/src/org/oscim/map/Animator.java]: It's a kinetic scroll somehow (probably not perfect), isn't it? I'd avoid breaking public API if possible.
BTW the "ease" word is not very recognizable, most won't understand what means, we can avoid it(?).

**Comment by devemux86** [comment]
We can have `Animator2` extending `Animator`.
Using inherited variables and methods, leaving public ones unmodified, overriding those needed changes.

**Comment by devemux86** [comment]
We can have `Animator2` extending `Animator`.
Using inherited variables and methods, leaving public ones unmodified, overriding those needed changes.

File: vtm/src/org/oscim/map/Animator.java (hunk 5 above)
Comment (Gustl22) [reply vtm/src/org/oscim/map/Animator.java]: > It's a kinetic scroll somehow (probably not perfect), isn't it? I'd avoid breaking public API if possible.

Ok, agreed.

> BTW the "ease" word is not very recognizable, most won't understand what means, we can avoid it(?).

Sure, but didn't find a better description for "ease" animations :| Any proposals?

Comment (devemux86) [reply vtm/src/org/oscim/map/Animator.java]: How about simply `kineticScroll` as the new `animateFling`?

**Comment by devemux86** [comment]
BTW (in this or any PR) if you wish to implement changes clean in a new PR instead of revert / move around this code, please do as you feel more easy to work. 🙂

File: vtm/src/org/oscim/map/Animator.java (hunk 5 above)
Comment (Gustl22) [reply vtm/src/org/oscim/map/Animator.java]: What `Fling` do you mean? The one with "ease" or the "physical" fling?
I wanted to keep calling physical as "fling" cause Android uses this as term.

Comment (devemux86) [reply vtm/src/org/oscim/map/Animator.java]: Meant to rename `animateEaseScroll` to `kineticScroll`, as there is already none "fling" in the name.

Comment (devemux86) [reply vtm/src/org/oscim/map/Animator.java]: Ah there is also the new `animateFlingScroll`, what's their difference?

Comment (devemux86) [reply vtm/src/org/oscim/map/Animator.java]: `animateEaseScroll` seems a changed `animateFling` and at the same time have a new `animateFlingScroll`.
So can keep old public `animateFling` as is and just add the new `animateFlingScroll` in `Animator2`?

**Comment by Gustl22** [comment]
> BTW (in this or any PR) if you wish to implement changes clean in a new PR instead of revert / move around this code, please do as you feel more easy to work. 🙂

Thanks, but wanted to keep issue in one PR.

Updated code. Feel free to rename or change while merging.

**Comment by devemux86** [comment]
What's the need of `Parameters.FLING`?
Only the `ANIMATOR2` seems needed and used.

**Comment by Gustl22** [comment]
Don't get it...simlpy use fling instead of kinetic, like described...

**Comment by devemux86** [comment]
All checks are:
```java
if (Parameters.FLING && mMap.animator() instanceof Animator2)
```
So selection between the 2 classes is sufficient, not need for something else(?).

**Comment by Gustl22** [comment]
You could use new implementation of kinetic (`kineticScroll` in Animator2, too, without flling...
Without param you can't try it... Is that what you meant?

**Comment by devemux86** [comment]
The `Parameters.FLING` is not used inside `Animator2`, only in the 3 similar [checks](https://github.com/mapsforge/vtm/pull/497#issuecomment-363780291).
Am I missing something?

**Comment by Gustl22** [comment]
Thought this shouldn't be decided inside `Animator2` class. Then the methods could have the same name, and only change behavior if some flag inside the class is set(?)...And can't decide if you want uncouple from current implementation in a specific case...

**Comment by devemux86** [comment]
```java
if (Parameters.ANIMATOR2)
```
This check seems sufficient in all cases, cannot see why need an extra flag in current implementation(?).

**Comment by Gustl22** [comment]
Cases:
- `ANIMATOR2 = false, FLING = false`: use current master kinetic implementation.
- `ANIMATOR2 = true, FLING = false`: use new implementation of kineticScroll. As you wasn't sure if new implementation - wich uses `moveTo` instead of `moveMap` - does work.
- `ANIMATOR2 = true, FLING = true`: use new physical fling (Animator2 needed).

**Comment by devemux86** [comment]
Things got confused..
If `ANIMATOR2` is `false` then only `Animator` class is created and current code is in action.
If `ANIMATOR2` is `true` then only `Animator2` class is created and new code is in action.
So extra flag seems unused.

**Comment by Gustl22** [comment]
Yes, but how can the user decide, that he wants to use `flingScroll` instead of `kineticScroll`?

**Comment by devemux86** [comment]
`animateFling` should remain only in `Animator` without override in `Animator2`.
Users select which implementation want to use / test, old or new, no need to mix them.

**Comment by devemux86** [comment]
> The first two ususally doesn't differ in behaviour...

Precisely, seen same behavior, so no need to complicate things.

File: vtm/src/org/oscim/map/Animator2.java [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                             int xmin, int xmax, int ymin, int ymax) {
+        kineticScroll(velocityX, velocityY, xmin, xmax, ymin, ymax);
+    }
+
+    /**
+     * @param velocityX the x velocity depends on screen resolution
+     * @param velocityY the y velocity depends on screen resolution
+     */
+    public void kineticScroll(float velocityX, float velocityY,
+                                  int xmin, int xmax, int ymin, int ymax) {
```
Comment (devemux86) [inline vtm/src/org/oscim/map/Animator2.java]: What's its difference with `Animator.animateFling`?
Cannot see any and some docs could help solve the confusion in posts. 🙂

**Comment by Gustl22** [comment]
Ok, should I update PR or is it a problem if you change code with your purpose?

**Comment by devemux86** [comment]
I'll do some tests and can update the code during merge.

File: vtm/src/org/oscim/map/Animator2.java (hunk 6 above)
Comment (Gustl22) [reply vtm/src/org/oscim/map/Animator2.java]: The difference is that it uses same sheme of the other ease animations (especially in `update`). The result keeps the same. Next time I'll do more docs then.

**Comment by devemux86** [comment]
Thanks for the nice work and keeping default implementation!
Squashed and merged via 03eb28f09ed3c1d158c5d2254246f10cb5c581e1.

I added also `Animator2` samples and the `Parameters` flag in first post.

**Comment by devemux86** [comment]
Suppose can enable the other flings (rotate, scale, etc) now in their own PR.

**Comment by boldtrn** [comment]
Good job @Gustl22. I tried it and it feels nice. :+1:

**Pull request closed without merging**